src/S3/S3Transfer/S3TransferManager.php

```diff
@@ -0,0 +1,850 @@
+        if (isset($config['filter'])) {
+            if (!is_callable($config['filter'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `filter` must be callable."
+                );
+            }
+
+            $filter = $config['filter'];
+        }
+
+        $uploadObjectRequestModifier = null;
+        if (isset($config['upload_object_request_modifier'])) {
+            if (!is_callable($config['upload_object_request_modifier'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `upload_object_request_modifier` must be callable."
+                );
+            }
+
+            $uploadObjectRequestModifier = $config['upload_object_request_modifier'];
+        }
+
+        $failurePolicyCallback = null;
+        if (isset($config['failure_policy'])) {
+            if (!is_callable($config['failure_policy'])) {
+                throw new InvalidArgumentException(
+                    "The provided config `failure_policy` must be callable."
+                );
+            }
+
+            $failurePolicyCallback = $config['failure_policy'];
+        }
+
+        $sourceDirectory = $uploadDirectoryRequest->getSourceDirectory();
+        $dirIterator = new RecursiveDirectoryIterator(
+            $sourceDirectory
+        );
+
+        $flags = FilesystemIterator::SKIP_DOTS;
+        if ($config['follow_symbolic_links'] ?? false) {
+            $flags |= FilesystemIterator::FOLLOW_SYMLINKS;
+        }
+
+        $dirIterator->setFlags($flags);
+
+        if ($config['recursive'] ?? false) {
+            $dirIterator = new RecursiveIteratorIterator(
+                $dirIterator,
+                RecursiveIteratorIterator::SELF_FIRST
+            );
+            if (isset($config['max_depth'])) {
+                if (!is_int($config['max_depth'])) {
+                    throw new InvalidArgumentException(
+                        "The provided config `max_depth` must be an integer."
+                    );
+                }
+
+                $dirIterator->setMaxDepth($config['max_depth']);
+            }
+        }
+
+        // To avoid circular symbolic links traversal
+        $dirVisited = [];
+        $files = filter(
+            $dirIterator,
+            function ($file) use ($filter, &$dirVisited) {
+                if (is_dir($file)) {
+                    $dirRealPath = realpath($file);
+                    if ($dirVisited[$dirRealPath] ?? false) {
+                        throw new RuntimeException(
+                            "A circular symbolic link traversal have been detected at $file -> $dirRealPath"
+                        );
+                    }
+                    $dirVisited[$dirRealPath] = true;
+                }
+
+                if ($filter !== null) {
+                    return !is_dir($file) && $filter($file);
+                }
+
+                return !is_dir($file);
+            }
+        );
```

Comment:
Shouldn't we only be doing this if `FilesystemIterator::FOLLOW_SYMLINKS` is enabled?

Reply:
I don`t think we need to scope this logic to just when symbolic links are enabled, since we need the filtering anyways.